src/include/column_inline.h

```diff
@@ -16,8 +16,15 @@ __col_insert_search_gt(WT_INSERT_HEAD *ins_head, uint64_t recno)
     WT_INSERT *ins, **insp, *ret_ins;
     int i;
 
+    /*
+     * CPU may reorder the reads and return a stale value of the tail.
+     *
+     * Place a read barrier here to avoid this issue.
```

Comment:
Do you need this one? It's not automatically a problem to see a stale value for the tail. (And in any case, another thread can concurrently write to the end, and we may not see it, whether or not there's a read barrier.) The problem seems more about ordering as you drop down levels, which I think is correctly handled further down.

Reply:
I think it is problem. Suppose the thread can actually see the insert list newly inserted to the new skip list, returning NULL here is an inconsistent read.

Reply:
I've discussed this IRL with Chenhao - we agree that this one might not be necessary. However, neither of us are 100% sure about leaving this without a barrier (e.g. some other code looking at the `NULL` we might erroneously return and expecting that to be consistent with some other view of the list). So I'll leave this thread open for discussion with @agorrod and @jeremythorp .

Reply:
This function is (currently) only called in one place. It is inlined, so any out-of-order reads could occur quite a long way in advance. However, in this case I *think* this barrier is not needed, but like @korteland and @quchenhao , I'm not certain yet. The barriers added later in the file are, I think, needed.